sql: panic during SHOW TRACE for SELECT COUNT(*)

Seen in #19199.

A user is seeing the following panic whenever they run the query `show trace for select count(*) from public.account`:

```
E171102 18:25:36.157731 17922 sql/session.go:1370  [client=[::1]:43328,user=root,n1] a SQL panic has occurred!
E171102 18:25:36.157781 17922 util/log/crash_reporting.go:117  [n1] a panic has occurred!
panic: decoding unset EncDatum [recovered]
	panic: panic while executing "show trace for select count(*) from public.account": decoding unset EncDatum [recovered]
	panic: panic while executing "show trace for select count(*) from public.account": decoding unset EncDatum [recovered]
	panic: panic while executing "show trace for select count(*) from public.account": decoding unset EncDatum

goroutine 17922 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc42067e120, 0x7fd9ea980140, 0xc42049f110)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:183 +0xb2
panic(0x1b44e00, 0xc42028c320)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc42065d500, 0x7fd9ea980140, 0xc420e214d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:423 +0xa1
panic(0x1b44e00, 0xc42028c320)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc42198a000, 0x1d08e47, 0x9, 0xc42198c048, 0x32)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1381 +0x28d
panic(0x1a890a0, 0xc420347e70)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*EncDatum).EnsureDecoded(0xc4219a58f8, 0xc4219a59d8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:175 +0x533
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.prettyEncDatums(0xc420be1420, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:300 +0x12a
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).processKV(0xc421104968, 0x7fd9ea922a70, 0xc420c82d40, 0xc420f11dc0, 0x17, 0x20, 0xc4213f4818, 0x1, 0x0, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:321 +0xdad
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).NextRow(0xc421104968, 0x7fd9ea922a70, 0xc420c82d40, 0x1, 0x1, 0x1, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:576 +0xa5
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).NextRowDecoded(0xc421104968, 0x7fd9ea922a70, 0xc420c82d40, 0x1, 0x0, 0x7fd9ea9c14b0, 0x2adeec0, 0xc4219a5f58, 0xc4219a5f58)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:598 +0x4c
github.com/cockroachdb/cockroach/pkg/sql.(*scanNode).Next(0xc421104840, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0xc4219a5fe8, 0x7fd9ea980140, 0xd299dc601c27820)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/scan.go:174 +0x89
github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).Next(0xc4213f2c60, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x1, 0x1c225e0, 0x19c9354)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:106 +0x56
github.com/cockroachdb/cockroach/pkg/sql.(*groupNode).Next(0xc4215588c0, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x0, 0xc4202e1100, 0x2ad6500)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/group.go:356 +0x834
github.com/cockroachdb/cockroach/pkg/sql.(*traceNode).Next.func1(0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0xc420071ce0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/trace.go:109 +0x131
github.com/cockroachdb/cockroach/pkg/sql.(*traceNode).Next(0xc420071ce0, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x7fd9ea9c14b0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/trace.go:129 +0xb1
github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).Next(0xc4213f2b40, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x1c9f420, 0x1, 0xc4205b7a80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:106 +0x56
github.com/cockroachdb/cockroach/pkg/sql.(*windowNode).Next(0xc4214e0280, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0xc42108e348, 0x0, 0xc4205b7a80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/window.go:473 +0x98
github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).Next(0xc4213f2a20, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x0, 0xc4219a6370, 0x1169bf9)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:106 +0x56
github.com/cockroachdb/cockroach/pkg/sql.(*windowNode).Next(0xc4214e0500, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x0, 0xc4205b7a80, 0xc4219a63e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/window.go:473 +0x98
github.com/cockroachdb/cockroach/pkg/sql.(*sortNode).Next(0xc420be0d90, 0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sort.go:453 +0xdd
github.com/cockroachdb/cockroach/pkg/sql.forEachRow(0x7fd9ea922a70, 0xc420c82d40, 0xc42198a208, 0x2acaf20, 0xc420be0d90, 0xc4219a6490, 0x0, 0x145a1ef)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2018 +0x53
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execClassic(0xc4206806c0, 0xc42198a208, 0x2acaf20, 0xc420be0d90, 0x2ad4ba0, 0xc42065d500, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2002 +0x2df
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4206806c0, 0x7fd9ea93cab0, 0xc42027a240, 0x0, 0x0, 0x0, 0x14f3581266b7e64f, 0x1, 0xc4219dfc20, 0xc42198a208, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2149 +0x637
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4206806c0, 0xc42198a000, 0x7fd9ea93cab0, 0xc42027a240, 0x0, 0x0, 0x0, 0x14f3581266b7e64f, 0x1, 0xc4219dfc20, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1788 +0x72b
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execSingleStatement(0xc4206806c0, 0xc42198a000, 0x7fd9ea93cab0, 0xc42027a240, 0x0, 0x0, 0x0, 0x14f3581266b7e64f, 0x1, 0xc4219dfc20, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1315 +0x92b
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4206806c0, 0xc42198a000, 0xc420c82bc0, 0x1, 0x1, 0x0, 0x1, 0xd70001, 0x0, 0x2ace780, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1192 +0x37f
github.com/cockroachdb/cockroach/pkg/sql.runWithAutoRetry(0xc4206806c0, 0xc42198a000, 0xc420c82bc0, 0x1, 0x1, 0x9400101, 0x0, 0x0, 0x3ff0000000000000, 0x2f74ba0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1016 +0x1f0
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed(0xc4206806c0, 0xc42198a000, 0xc420c82bc0, 0x1, 0x1, 0x0, 0x0, 0xc42001b300, 0x60a1d3)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:834 +0x26a
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4206806c0, 0xc42198a000, 0xc42198c048, 0x32, 0x0, 0x0, 0x32, 0xfb8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:757 +0x493
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4206806c0, 0xc42198a000, 0xc42198c048, 0x32, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:638 +0x129
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc42065d500, 0xc42065d528, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:612 +0x148
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc42065d500, 0x7fd9ea980140, 0xc420e214d0, 0xc4202bb8f0, 0x5400, 0xc4200b2168, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:503 +0xbd0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4200b2000, 0x7fd9ea922a70, 0xc420425f40, 0x2ad4d80, 0xc421558320, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x96c
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x2ad4d80, 0xc421558320)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:748 +0x149
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc42067e120, 0x7fd9ea980140, 0xc42049f110, 0xc42081e008, 0x2ad4d80, 0xc421558320, 0xc420e311a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:140 +0xa9
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0x267
```

Assigning Radu, but feel free to reassign.

--- a/pkg/sql/sqlbase/rowfetcher.go
+++ b/pkg/sql/sqlbase/rowfetcher.go
@@ -318,7 +318,11 @@ func (rf *RowFetcher) processKV(
 	ctx context.Context, kv client.KeyValue, debugStrings bool,
 ) (prettyKey string, prettyValue string, err error) {
 	if debugStrings {
-		prettyKey = fmt.Sprintf("/%s/%s%s", rf.desc.Name, rf.index.Name, prettyEncDatums(rf.keyVals))
+		if rf.mustDecodeIndexKey {
+			prettyKey = fmt.Sprintf("/%s/%s%s", rf.desc.Name, rf.index.Name, prettyEncDatums(rf.keyVals))
+		} else {
+			prettyKey = fmt.Sprintf("/%s/%s/{not-decoded}", rf.desc.Name, rf.index.Name)
+		}
 	}
 
 	if rf.indexKey == nil {

--- a/pkg/sql/trace_test.go
+++ b/pkg/sql/trace_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 )
@@ -330,4 +331,19 @@ func TestBracketInTracetags(t *testing.T) {
 	if !ok {
 		t.Fatal("expected message not found in trace")
 	}
+}
+
+func TestKVTraceWithCountStar(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Test that we don't crash if we try to do a KV trace
+	// on a COUNT(*) query (#19846).
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(context.TODO())
+
+	r := sqlutils.MakeSQLRunner(t, db)
+	r.Exec("CREATE DATABASE test")
+	r.Exec("CREATE TABLE test.a (a INT PRIMARY KEY, b INT)")
+	r.Exec("INSERT INTO test.a VALUES (1,1), (2,2)")
+	r.Exec("SHOW KV TRACE FOR SELECT COUNT(*) FROM test.a")
 }

